# Merge Stockwatcher with CRSP

In [1]:
import numpy as np
import pandas as pd
from datetime import datetime

In [2]:
pd.set_option('display.max_columns', None)

In [3]:
sw_house = pd.read_csv("../data/house_all_transactions_fixed.csv")
sw_senate = pd.read_csv("../data/senate_all_transactions.csv")
sw_senate["transaction_date"] = list(map(lambda x: datetime.strptime(x, "%m/%d/%Y").strftime("%Y-%m-%d"), sw_senate["transaction_date"]))

crsp_house = pd.read_csv("../data/house_crsp.zip")
crsp_senate = pd.read_csv("../data/senate_crsp.zip")

## House

In [4]:
sw_house.shape

(15432, 12)

In [5]:
sw_crsp_house = sw_house.merge(crsp_house, how="left", left_on=["transaction_date", "ticker"], right_on=["dlycaldt", "ticker"])
sw_crsp_house

disclosure_year disclosure_date transaction_date  owner ticker  \
0                 2021      10/04/2021       2021-09-27  joint     BP   
1                 2021      10/04/2021       2021-09-13  joint    XOM   
2                 2021      10/04/2021       2021-09-10  joint   ILPT   
3                 2021      10/04/2021       2021-09-28  joint     PM   
4                 2021      10/04/2021       2021-09-17   self    BLK   
...                ...             ...              ...    ...    ...   
15480             2020      06/10/2020       2020-04-09     --    SWK   
15481             2020      06/10/2020       2020-04-09     --    USB   
15482             2020      06/10/2020       2020-03-13    NaN    BMY   
15483             2020      06/10/2020       2020-03-13    NaN    LLY   
15484             2020      06/10/2020       2020-03-13    NaN    DIS   

                                       asset_description          type  \
0                                                 BP plc      purchase   
1                                Exxon Mobil Corporation      purchase   
2      Industrial Logistics Properties Trust - Common...      purchase   
3                       Phillip Morris International Inc      purchase   
4                                          BlackRock Inc  sale_partial   
...                                                  ...           ...   
15480                       Stanley Black & Decker, Inc.  sale_partial   
15481                                       U.S. Bancorp  sale_partial   
15482                       Bristol-Myers Squibb Company     sale_full   
15483                              Eli Lilly and Company     sale_full   
15484                                Walt Disney Company     sale_full   

                      amount            representative district  \
0           $1,001 - $15,000        Hon. Virginia Foxx     NC05   
1           $1,001 - $15,000        Hon. Virginia Foxx     NC05   
2          $15,001 - $50,000        Hon. Virginia Foxx     NC05   
3          $15,001 - $50,000        Hon. Virginia Foxx     NC05   
4           $1,001 - $15,000    Hon. Alan S. Lowenthal     CA47   
...                      ...                       ...      ...   
15480       $1,001 - $15,000        Hon. Ed Perlmutter     CO07   
15481       $1,001 - $15,000        Hon. Ed Perlmutter     CO07   
15482    $100,001 - $250,000  Hon. Nicholas Van Taylor     TX03   
15483  $500,001 - $1,000,000  Hon. Nicholas Van Taylor     TX03   
15484    $250,001 - $500,000  Hon. Nicholas Van Taylor     TX03   

                                                ptr_link  \
0      https://disclosures-clerk.house.gov/public_dis...   
1      https://disclosures-clerk.house.gov/public_dis...   
2      https://disclosures-clerk.house.gov/public_dis...   
3      https://disclosures-clerk.house.gov/public_dis...   
4      https://disclosures-clerk.house.gov/public_dis...   
...                                                  ...   
15480  https://disclosures-clerk.house.gov/public_dis...   
15481  https://disclosures-clerk.house.gov/public_dis...   
15482  https://disclosures-clerk.house.gov/public_dis...   
15483  https://disclosures-clerk.house.gov/public_dis...   
15484  https://disclosures-clerk.house.gov/public_dis...   

       cap_gains_over_200_usd   permno secinfostartdt secinfoenddt  \
0                       False  29890.0     2009-03-27   2022-06-30   
1                       False  11850.0     2019-09-11   2022-06-30   
2                       False  17222.0     2018-03-23   2022-06-30   
3                       False  92602.0     2012-05-07   2022-06-30   
4                       False  87267.0     2019-11-20   2022-06-30   
...                       ...      ...            ...          ...   
15480                   False  43350.0     2020-11-03   2022-06-30   
15481                   False  66157.0     2020-04-06   2022-06-30   
15482                   False  19393.0     2008-06-27   2022-06-30   
15483          

In [6]:
sw_crsp_house.dropna(subset=["dlyprc"]).shape

(12175, 105)

In [7]:
sw_crsp_house.dropna(subset=["dlyprc"]).shape[0] / sw_house.replace("--", np.nan).dropna(subset=["ticker"]).shape[0]

0.8593915437283829

In [8]:
# sw_crsp_house.to_csv("../data/sw_crsp_house.csv", index=False)

In [9]:
compression_opts = dict(method='zip', archive_name='sw_crsp_house.csv')  
sw_crsp_house.to_csv('sw_crsp_house.zip', index=False,compression=compression_opts)

In [10]:
%%bash
mv sw_crsp_house.zip ../data/sw_crsp_house.zip

## Senate

In [11]:
sw_senate.shape

(9553, 11)

In [12]:
sw_crsp_senate = sw_senate.merge(crsp_senate, how="left", left_on=["transaction_date", "ticker"], right_on=["dlycaldt", "ticker"])
sw_crsp_senate

transaction_date   owner ticker  \
0          2022-08-10  Spouse     --   
1          2022-08-01    Self    IVV   
2          2022-08-01    Self   EFAV   
3          2022-08-01    Self    XLY   
4          2022-08-01    Self   VTRS   
...               ...     ...    ...   
9659       2012-08-17     NaN    NaN   
9660       2012-08-16     NaN    NaN   
9661       2012-08-15     NaN    NaN   
9662       2012-08-02     NaN    NaN   
9663       2012-07-25     NaN    NaN   

                                      asset_description            asset_type  \
0     Parker-Hannifin Corp 701094AQ7 <div class="tex...        Corporate Bond   
1                          iShares Core S&amp;P 500 ETF                 Stock   
2                  iShares MSCI EAFE Min Vol Factor ETF                 Stock   
3      SPDR Select Sector Fund - Consumer Discretionary                 Stock   
4                           Viatris Inc. - Common Stock                 Stock   
...                                                 ...                   ...   
9659  This filing was disclosed via scanned PDF. Use...  PDF Disclosed Filing   
9660  This filing was disclosed via scanned PDF. Use...  PDF Disclosed Filing   
9661  This filing was disclosed via scanned PDF. Use...  PDF Disclosed Filing   
9662  This filing was disclosed via scanned PDF. Use...  PDF Disclosed Filing   
9663  This filing was disclosed via scanned PDF. Use...  PDF Disclosed Filing   

             type             amount comment              senator  \
0        Purchase  $15,001 - $50,000      --      Susan M Collins   
1     Sale (Full)  $15,001 - $50,000      --  Thomas H Tuberville   
2     Sale (Full)  $15,001 - $50,000      --  Thomas H Tuberville   
3     Sale (Full)  $15,001 - $50,000      --  Thomas H Tuberville   
4     Sale (Full)   $1,001 - $15,000      --  Thomas H Tuberville   
...           ...                ...     ...                  ...   
9659          NaN            Unknown     NaN   Sheldon Whitehouse   
9660          NaN            Unknown     NaN          Pat Roberts   
9661          NaN            Unknown     NaN     Robert J Portman   
9662          NaN            Unknown     NaN      Thomas R Carper   
9663          NaN            Unknown     NaN    Benjamin L Cardin   

                                               ptr_link disclosure_date  \
0     https://efdsearch.senate.gov/search/view/ptr/7...      09/16/2022   
1     https://efdsearch.senate.gov/search/view/ptr/7...      09/15/2022   
2     https://efdsearch.senate.gov/search/view/ptr/7...      09/15/2022   
3     https://efdsearch.senate.gov/search/view/ptr/7...      09/15/2022   
4     https://efdsearch.senate.gov/search/view/ptr/7...      09/15/2022   
...                                                 ...             ...   
9659  https://efdsearch.senate.gov/search/view/paper...      08/17/2012   
9660  https://efdsearch.senate.gov/search/view/paper...      08/16/2012   
9661  https://efdsearch.senate.gov/search/view/paper...      08/15/2012   
9662  https://efdsearch.senate.gov/search/view/paper...      08/02/2012   
9663  https://efdsearch.senate.gov/search/view/paper...      07/25/2012   

      permno secinfostartdt secinfoenddt securitybegdt securityenddt  \
0        NaN            NaN          NaN           NaN           NaN   
1        NaN            NaN          NaN           NaN           NaN   
2        NaN            NaN          NaN           NaN           NaN   
3        NaN            NaN          NaN           NaN           NaN   
4        NaN            NaN          NaN           NaN           NaN   
...      ...            ...          ...           ...           ...   
9659     NaN            NaN          NaN           NaN           NaN   
9660     NaN            NaN          NaN           NaN           NaN   
9661     NaN            NaN          NaN           NaN           NaN   
9662     NaN            NaN          NaN           NaN           NaN   
9663     NaN            NaN  

In [13]:
sw_crsp_senate.dropna(subset=["dlyprc"]).shape

(6536, 104)

In [14]:
sw_crsp_senate.dropna(subset=["dlyprc"]).shape[0] / sw_senate.replace("--", np.nan).dropna(subset=["ticker"]).shape[0]

0.9094197857242243

In [15]:
# sw_crsp_senate.to_csv("../data/sw_crsp_senate.csv", index=False)

In [16]:
compression_opts = dict(method='zip', archive_name='sw_crsp_senate.csv')  
sw_crsp_senate.to_csv('sw_crsp_senate.zip', index=False,compression=compression_opts)

In [17]:
%%bash
mv sw_crsp_senate.zip ../data/sw_crsp_senate.zip

#### Coverage Report:
- 85.9\% of house stockwatcher data have corresponding CRSP available. 
- 90.9\% of senate stockwatcher data have corresponding CRSP available. 